diff --git a/pdf_viewer/input.cpp b/pdf_viewer/input.cpp
index 44aa89a9..1c8003bb 100644
--- a/pdf_viewer/input.cpp
+++ b/pdf_viewer/input.cpp
@@ -5260,7 +5260,7 @@ public:
     void perform() {
         std::optional<Path> key_file_path = widget->input_handler->get_or_create_user_keys_path();
         if (key_file_path) {
-            open_file(key_file_path.value().get_path(), true);
+            open_text_file(key_file_path.value().get_path(), true);
         }
     }
 
@@ -5287,7 +5287,7 @@ public:
     PrefsCommand(MainWidget* w) : Command(cname, w) {};
 
     void perform() {
-        open_file(default_config_path.get_path(), true);
+        open_text_file(default_config_path.get_path(), true);
     }
 
     bool requires_document() { return false; }
@@ -5302,7 +5302,7 @@ public:
     void perform() {
         std::optional<Path> pref_file_path = widget->config_manager->get_or_create_user_config_file();
         if (pref_file_path) {
-            open_file(pref_file_path.value().get_path(), true);
+            open_text_file(pref_file_path.value().get_path(), true);
         }
     }
 
diff --git a/pdf_viewer/main_widget.cpp b/pdf_viewer/main_widget.cpp
index 2b82cc20..9c6d93bb 100644
--- a/pdf_viewer/main_widget.cpp
+++ b/pdf_viewer/main_widget.cpp
@@ -6783,7 +6783,7 @@ void MainWidget::handle_keys_user_all() {
         keys_paths_wstring,
         [&](std::wstring* path) {
             if (path) {
-                open_file(*path, true);
+                open_text_file(*path, true);
             }
         },
         this));
@@ -6803,7 +6803,7 @@ void MainWidget::handle_prefs_user_all() {
         prefs_paths_wstring,
         [&](std::wstring* path) {
             if (path) {
-                open_file(*path, true);
+                open_text_file(*path, true);
             }
         },
         this));
diff --git a/pdf_viewer/utils.cpp b/pdf_viewer/utils.cpp
index 785d5fd3..33848bd1 100644
--- a/pdf_viewer/utils.cpp
+++ b/pdf_viewer/utils.cpp
@@ -1285,6 +1285,25 @@ void open_file(const std::wstring& path, bool show_fail_message) {
 
 }
 
+void open_text_file(const std::wstring& path, bool show_fail_message) {
+    QString editor = qEnvironmentVariable("VISUAL");
+    if (editor.isEmpty()) {
+        editor = qEnvironmentVariable("EDITOR");
+    }
+
+    QString canonical_path = QString::fromStdWString(get_canonical_path(path));
+    QStringList arguments = QProcess::splitCommand(editor);
+    if (!arguments.isEmpty()) {
+        QString program = arguments.takeFirst();
+        arguments.push_back(canonical_path);
+        if (QProcess::startDetached(program, arguments)) {
+            return;
+        }
+    }
+
+    open_file_url(canonical_path, show_fail_message);
+}
+
 void get_text_from_flat_chars(const std::vector<fz_stext_char*>& flat_chars, std::wstring& string_res, std::vector<int>& indices) {
 
     string_res.clear();
diff --git a/pdf_viewer/utils.h b/pdf_viewer/utils.h
index 0cb30b6c..4b2014e5 100644
--- a/pdf_viewer/utils.h
+++ b/pdf_viewer/utils.h
@@ -91,6 +91,7 @@ void open_file_url(const std::wstring& file_url, bool show_fail_message);
 void open_web_url(const std::wstring& web_url);
 
 void open_file(const std::wstring& path, bool show_fail_message);
+void open_text_file(const std::wstring& path, bool show_fail_message);
 void search_custom_engine(const std::wstring& search_string, const std::wstring& custom_engine_url);
 void search_google_scholar(const std::wstring& search_string);
 void search_libgen(const std::wstring& search_string);
